On-hover animation for node and link colors

mpryor opened the pull request:
Starting a draft PR to play around with some animation concepts

jollyjerr reviewed (commented):
Something I noticed while playing around with this - we will want to update all our interactions to restore 60fps or whatever the simulation runs at. Right now this looks awesome while the simulation is running, but once it's over this animation drops to a really slow frame rate.

https://github.com/jollyjerr/mindgraph/assets/51095001/2ffbb845-bf6c-4348-a805-d4286b4ba671

mpryor commented:
@jollyjerr  Yeah for sure!! I've been wondering about how we want to handle our core-update loop in general. The only reason this even animates after the simulation is done is because I added that random `setInterval()` in, which may not be ideal.

Some things to think about:
1. Do we want to just have a statically defined refresh interval? (Probably 1/60, for example), and just consistently call re-renders on that interval? 
2. Or do we want a dynamic delta between frames based off the elapsed time from the previous frame?
3. Do we only want to call redraws when things are worth redrawing? For example, if there's no simulation running, no animations running, and no user events (hover, click, etc) it might be silly to be redrawing the canvas

mpryor commented:
In https://github.com/jollyjerr/mindgraph/pull/19/commits/f1ff1b77d5cb1809f9cd51e78f4a9fcdc5baf391 I fixed the easing functions so they actually work now, added a couple more easing functions, and added an example page at http://localhost:5173/tests/animexample.html

jollyjerr commented:
IMHO pinning the whole thing to 60fps or https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame would be fine for now!

mpryor commented:
I updated the core rendering loop to use the `window.requestAnimationFrame()` per your suggestion and also actually working my way through Mozilla's Canvas tutorial it seems to be the preferred method.

Now the the rendering loop is controlled on an interval (outside the Artist class), any re-renders that were triggered internally can be removed (during zoom, during mouseover events, etc).

jollyjerr commented on packages/draw/src/animation.ts:
```diff
@@ -0,0 +1,123 @@
+      case 'easeout':
+        easingFunc = easeOut;
+        break;
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-case-declarations
+        const __exhaustiveCase: never = this.easing;
+        throw new Error('Not all easing functions are handled');
```
Default is already never in this case, so I think I'd recommend just doing something like this

```suggestion
      default:
        throw new Error(`Easing function ${this.easing} not handled!`);
```

<img width="1170" alt="Screenshot 2023-06-19 at 10 27 47 AM" src="https://github.com/jollyjerr/mindgraph/assets/51095001/ce859091-3c18-44c3-ae89-c54124fbbf15">

mpryor replied:
@jollyjerr The `never` assignment provides compile type error checking for cases that aren't being handled (instead of runtime checking). In other words, if a new union was added to the `easing` type, the build will fail rather than a failure at runtime. 

See this: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking for a better explanation of what's going on. I added the `throw new Error` just for a "just in case this gets passed a build" scenario

mpryor commented on packages/draw/src/artist.ts:
```diff
@@ -76,7 +75,7 @@ export class Artist {
         minZoom: this.styles.minZoom,
         maxZoom: this.styles.maxZoom,
-        observers: [() => this.redraw()],
+        observers: [],
```
Done!

mpryor commented on packages/draw/src/renderables.ts:
```diff
@@ -4,18 +4,25 @@ import { ConfiguredSimulationLink } from './simulation';
   return val >= min && val <= max;
 }
 
+const ANIMATION_TIME = 0.3;
```
Done!
